system-addon/lib/PlacesFeed.jsm

```diff
@@ -0,0 +1,204 @@
+  }
+}
+
+/**
+ * HistoryObserver - observes events from PlacesUtils.history
+ */
+class HistoryObserver extends Observer {
+  constructor(dispatch) {
+    super(dispatch, Ci.nsINavHistoryObserver);
+  }
+
+  /**
+   * onDeleteURI - Called when an link is deleted from history.
+   *
+   * @param  {obj} uri        A URI object representing the link's url
+   *         {str} uri.spec   The URI as a string
+   */
+  onDeleteURI(uri) {
+    this.dispatch({
+      type: at.PLACES_LINK_DELETED,
+      data: {url: uri.spec}
+    });
+  }
+
+  /**
+   * onClearHistory - Called when the user clears their entire history.
+   */
+  onClearHistory() {
+    this.dispatch({type: at.PLACES_HISTORY_CLEARED});
+  }
+}
+
+/**
+ * BookmarksObserver - observes events from PlacesUtils.bookmarks
+ */
+class BookmarksObserver extends Observer {
+  constructor(dispatch) {
+    super(dispatch, Ci.nsINavBookmarkObserver);
+  }
+
+  /**
+   * onItemAdded - Called when a bookmark is added
+   *
+   * @param  {str} id
+   * @param  {str} folderId
+   * @param  {int} index
+   * @param  {int} type       Indicates if the bookmark is an actual bookmark,
+   *                          a folder, or a separator.
+   * @param  {str} uri
+   * @param  {str} title
+   * @param  {int} dateAdded
+   * @param  {str} guid      The unique id of the bookmark
+   */
+  async onItemAdded(...args) {
+    const type = args[3];
+    const guid = args[7];
+    if (type !== PlacesUtils.bookmarks.TYPE_BOOKMARK) {
+      return;
+    }
+    try {
+      // bookmark: {bookmarkGuid, bookmarkTitle, lastModified, url}
+      const bookmark = await NewTabUtils.activityStreamProvider.getBookmark(guid);
+      this.dispatch({type: at.PLACES_BOOKMARK_ADDED, data: bookmark});
+    } catch (e) {
+      Cu.reportError(e);
+    }
+  }
+
+  onItemRemoved(guid, folderId, index, type, uri) {
+    if (type === PlacesUtils.bookmarks.TYPE_BOOKMARK) {
+      this.dispatch({
+        type: at.PLACES_BOOKMARK_REMOVED,
+        data: {url: uri.spec, bookmarkGuid: guid}
+      });
+    }
+  }
+
+  /**
+   * onItemChanged - Called when a bookmark is modified
+   *
+   * @param  {str} id           description
+   * @param  {str} property     The property that was modified (e.g. uri, title)
+   * @param  {bool} isAnnotation
+   * @param  {any} value
+   * @param  {int} lastModified
+   * @param  {int} type         Indicates if the bookmark is an actual bookmark,
+   *                             a folder, or a separator.
+   * @param  {int} parent
+   * @param  {str} guid         The unique id of the bookmark
+   */
+  async onItemChanged(...args) {
+    const property = args[1];
+    const type = args[5];
+    const guid = args[7];
+
+    // Only process this event if it is a TYPE_BOOKMARK, and uri or title was the property changed.
+    if (type !== PlacesUtils.bookmarks.TYPE_BOOKMARK || !["uri", "title"].includes(property)) {
```

Comment:
does changing the uri or the title change the frecency? is that why we only care about those?